### PR TITLE
Token allowance flow update 

### PR DESCRIPTION
**Pull request by NiranjanaBinoy**
## Explanation
Fixes [#610](https://github.com/MetaMask/MetaMask-planning/issues/610)
### Changes introduced in this PR:

-  Updated the copy `Custom spending cap`   to `Site requested spending cap` when
-  The copies `Set a spending cap for your` and `Review the spending cap for your`are updated to `Spending cap request for your`
- The copy `Use default` is updated to  `Use site suggestion`
- The spending cap on the first page of Approve screen is pre-filled with the dapp suggested spending cap instead of having a blank input field.
-  Learn More link is added below the copy. The link is directing to https://support.metamask.io/hc/en-us/articles/6055177143579-How-to-customize-token-approvals-with-a-spending-cap.

## Screenshots/Screencaps

https://github.com/MetaMask/metamask-extension/assets/43930900/faf103e9-aff0-4178-9200-5072ffab11e6

## Manual Testing Steps
1. Create a new token from the dapp
2. Click the `Approve Token` button and make sure the token allowance page loads prefilled with the spending cap.
3. Make sure all the labels are updated based on the description.

## Pre-merge author checklist

- [ ] I've clearly explained:
  - [ ] What problem this PR is solving
  - [ ] How this problem was solved
  - [ ] How reviewers can test my changes
- [ ] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [ ] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

### CODE REVIEW

**Comment by digiwand** [comment]
Screen Shot 2023-06-20 at 5 33 18 PM

For some reason, the text uses 2 lines on my screen. Perhaps, has something changed since the video above was taken?

File: ui/pages/token-allowance/token-allowance.js [hunk 1]
```diff
@@ -122,24 +125,25 @@ export default function TokenAllowance({
   const unapprovedTxCount = useSelector(getUnapprovedTxCount);
   const unapprovedTxs = useSelector(getUnapprovedTransactions);
   const useCurrencyRateCheck = useSelector(getUseCurrencyRateCheck);
-  let customTokenAmount = useSelector(getCustomTokenAmount);
-  if (thisOriginIsAllowedToSkipFirstPage && dappProposedTokenAmount) {
-    customTokenAmount = dappProposedTokenAmount;
+  const customTokenAmount = useSelector(getCustomTokenAmount);
+  if (!thisOriginIsAllowedToSkipFirstPage && !dappProposedTokenAmount) {
+    // customTokenAmount = dappProposedTokenAmount;
```
Comment (digiwand) [inline ui/pages/token-allowance/token-allowance.js:130]:
```suggestion
```

File: ui/pages/token-allowance/token-allowance.js [hunk 2]
```diff
@@ -101,6 +101,9 @@ export default function TokenAllowance({
   const { hostname } = new URL(origin);
   const thisOriginIsAllowedToSkipFirstPage = ALLOWED_HOSTS.includes(hostname);
 
+  const [customTokenValue, setCustomTokenValue] = useState(
+    dappProposedTokenAmount,
+  );
```
Comment (digiwand) [inline ui/pages/token-allowance/token-allowance.js:106]: hi @NiranjanaBinoy, I'm having trouble differentiating between `customTokenValue` and `customTokenAmount`. Can you please explain?

File: ui/pages/token-allowance/token-allowance.js [hunk 3]
```diff
@@ -122,24 +125,25 @@ export default function TokenAllowance({
   const unapprovedTxCount = useSelector(getUnapprovedTxCount);
   const unapprovedTxs = useSelector(getUnapprovedTransactions);
   const useCurrencyRateCheck = useSelector(getUseCurrencyRateCheck);
-  let customTokenAmount = useSelector(getCustomTokenAmount);
-  if (thisOriginIsAllowedToSkipFirstPage && dappProposedTokenAmount) {
-    customTokenAmount = dappProposedTokenAmount;
+  const customTokenAmount = useSelector(getCustomTokenAmount);
+  if (!thisOriginIsAllowedToSkipFirstPage && !dappProposedTokenAmount) {
```
Comment (digiwand) [inline ui/pages/token-allowance/token-allowance.js:129]: this conditional is not the exact opposite of the former. was this intentional?

File: ui/pages/token-allowance/token-allowance.js (hunk 2 above)
Comment (NiranjanaBinoy) [reply ui/pages/token-allowance/token-allowance.js:106]: Sorry for the confusion, `customTokenValue` was the state variable that holds the custom value we entered for the spending cap, and `customTokenAmount` is the value from the state. I have renamed `customTokenValue` to `customSpendingCap` for more clarity.

File: ui/pages/token-allowance/token-allowance.js (hunk 3 above)
Comment (NiranjanaBinoy) [reply ui/pages/token-allowance/token-allowance.js]: Thanks for bringing this up ❤️ With the current changes, we default to dappProposedTokenAmount when we load the token allowance page. That is the reason I changed the previous conditional check. But your comment made me think, and I need to do more testing to decide whether the current condition I added is relevant.

**Comment by NiranjanaBinoy** [comment]
> Screen Shot 2023-06-20 at 5 33 18 PM
> For some reason, the text uses 2 lines on my screen. Perhaps, has something changed since the video above was taken?

Screenshot 2023-06-20 at 11 12 46 PM


It is appearing  correctly for me; could it be a resolution difference?

**Comment by seaona** [comment]
hi @NiranjanaBinoy !! The test failure seems legit. There is a Big number error whenever we update the amount value. 2 ways of getting this:

1. Update the amount value from the MetaMask full window

https://github.com/MetaMask/metamask-extension/assets/54408225/c61a2871-191e-40e3-859b-c49895039206

2. Update the amount value from the popup window


https://github.com/MetaMask/metamask-extension/assets/54408225/ea3e48eb-8eac-46c8-b340-a13bc1e89686

**Review by segun — DISMISSED** [review]
e2e is failing on chrome...otherwise looks good.

**Review by digiwand — DISMISSED** [review]

**Review by segun — DISMISSED** [review]

**Comment by seaona** [comment]
Overall the PR looks good!! 

## e2e failing
I can see how a Big Number error appears on the console whenever we open the Approve screen. This needs to be solved on the UI code level @NiranjanaBinoy as the e2e test expects no errors on the console.

image

image

## Metrics

I have one question around metrics, which are not clear to me.

About the approval amount vs balance ratio, I understand that if I have 10 tokens and I approve 10, this value should be 1 / 100%, but I see this is not the case. It seems that it's taking the dapp suggested value over the balance, but not the value that is finally set by the user. Is this expected? 

@bschorchit maybe this could be solved / further investigate in the Metrics issue instead of here, but just wanted to make sure.

| case   | user balance  |  site suggested value   | value confirmed  |  ` transaction_approval _amount_vs_balance_ratio: `  |  `transaction_approval _amount_vs_proposed_ratio: ` |
|---|---|---|---|---|---|
| same suggested and amount  | 10  |  7  | 7  | 70  | 100  |
| different suggested and amount |  10  | 7  | 10  | 70  | 70  |
| different suggested and amount  |  10  | 7  | 1  | 70  | 700 |
| 0 amount  |  10  | 7  | 0  | null  | null |

**Comment by bschorchit** [comment]
Thank you, @seaona! I noticed these too recently when investigating these metrics on Mixpanel and looking into the code for it. I think we're passing the wrong parameter to calculate it. I haven't opened an issue for this yet, but will do it soon. Let's keep this out of the scope of this PR for now.

**Comment by seaona** [comment]
thank you @bschorchit !!

**Comment by NiranjanaBinoy** [comment]
Hi @seaona , The BigNumber error seems to exist in the `develop` as well; not sure why it started giving us e2e failure all of a sudden.
The error was happening when we pass an empty string to `New BigNumber()` while calculating the gas estimate each time the spending cap is changed.

**Comment by seaona** [comment]
@NiranjanaBinoy great work!! I can see the issue fixed and everything looks good from QA :fire:

**Review by segun — APPROVED** [review]

**Review by digiwand — APPROVED** [review]

**Pull request merged**